neo4j/_async/work/session.py

```diff
@@ -278,9 +280,15 @@ async def job(tx, **job_kwargs):
                 return QueryResult([], summary)
             return await tx.query(query, parameters, **job_kwargs)
 
-        return await self.execute(job, **kwargs)
+        return await self.execute(
+            job,
+            cluster_member_access=cluster_member_access,
+            **kwargs
+        )
 
-    async def execute(self, transaction_function, *args, **kwargs):
+    async def execute(self, transaction_function, *args,
+                      cluster_member_access=CLUSTER_AUTO_ACCESS,
+                      **kwargs):
```

Comment:
2. I don't like this design. Here is the problem: if you write a wrapper function that does not control the signature of `transaction_function`, you are screwed, because you can't make it right in the general case because `transaction_function` could have a keyword argument `cluster_member_access`. What now?

This could be addressed like so
```python
    async def execute(
        self, transaction_function,
        transaction_function_args=None, transaction_function_kwargs=None,
        cluster_member_access=CLUSTER_AUTO_ACCESS
    ):
```
which is admittedly less nice to use. There is the third option
```python
    async def execute(
        self, transaction_function, *args,
        transaction_function_args=None, transaction_function_kwargs=None,
        cluster_member_access=CLUSTER_AUTO_ACCESS, **kwargs
    ):
```
But I'm not sure if that's not just more confusing than it's helping. I really don't know. I think this is the major downside of cramming all our API into easy to access functions... They become bloated, more complicated, and have to do many things at once.

Reply:
The `cluster_member_access` was already being not passed to the tx function in the previous version and it is kind of useless in that context, since none of the tx methods accept this argument.


I've tried to keep usage quite simple while keep the usage quite similar to the previous tx function. The difference is now the param which will be removed is documented.

An alternative is removing the extra arguments (*args and *kwargs), this way the tx function will only receive the transaction.

Reply:
I don't understand. Isn't `transaction_function` a function defined by user-level code (outside of the driver)? How can you know it doesn't have a `cluster_member_access` parameter?

Reply:
We don't know. The `Session.run` also have the same problem, the user couldn't have a parameter called `parameters` (if they set the param like `session.run('return n, parameters', n=1, parameters='a')`).

We could also always add this name params to the `tx_func` kwargs.

Reply:
While this is true about the `session.run`, if I was to write a wrapper, I'd just never call `session.run(query, **parameters)` but instead `session.run(query, parameters=parameters)`. That way, I can be sure that my wrapper can cope with any query parameters thrown at it.

> We could also always add this name params to the tx_func kwargs. Which is kind of the approach used by to extract metadata and timeout in read_transaction.

I can't follow, sorry.

Reply:
`read_transaction` and `write_transaction` get `timeout` and `metadata` from the `**kwargs` without remove it from it. We could do it with the `cluster_member_access` (and `database`, in the `driver.execute` case) by put it back to the `**kwargs`. However, we will have the issue with the documentation of the params. 

For solving this, I'm suggesting keep the signature like:

```
async def execute(
        self, transaction_function, *args,
        cluster_member_access=None, **kwargs
    ):
```

Then, in the execute implementation do something like:

```python
if cluster_member_access is None:
    cluster_member_access = CLUSTER_AUTO_ACCESS
else:
    kwargs['cluster_member_access'] = cluster_member_access
```

The problem is: we don't know if the user set named param `cluster_member_access` or not, we only can guess if they are using the default value.

Reply:
There is still another issue: the user could use a different meaning and dataset for the `cluster_member_access` variable.